addons/test_website/tests/test_redirect.py

```diff
@@ -231,3 +232,91 @@ def test_05_redirect_404_notfound_record(self):
+            'name': 'Test Portal Record',
+            'partner_id': cls.customer.id,
+            'user_id': cls.user_admin.id,
+        })
+        for group_name, group_func, group_data in cls.record_portal.sudo()._notify_get_recipients_groups(
+            cls.env['mail.message'], False
+        ):
+            if group_name == 'portal_customer' and group_func(cls.customer):
+                cls.record_access_url = group_data['button_access']['url']
+                break
+        else:
+            raise AssertionError('Record access URL not found')
+        cls.website = cls.env.ref('website.default_website')
+
+    def preload_res_lang_get_available(self):
+        """ Preload get_available language for default website and no website (see url_lang called by url_for) """
+        with MockRequest(self.env, website=self.website) as request:
+            request.is_front_end = True
+            request.env['res.lang'].with_context(website_id=self.website.id).sudo().get_available()
+            request.env['res.lang'].sudo().get_available()
+
+    def setup_website_customer_lang(self, lang_code):
+        """ Activate language for the website and set it to the customer. """
+        ResLang = self.env['res.lang']
+        lang = ResLang._activate_lang(lang_code) or ResLang._create_lang(lang_code)
+        self.website.language_ids += lang
+        self.customer.lang = lang_code
+        return lang
+
+    @mute_logger('odoo.addons.base.models.ir_model')
+    def test_customer_access_not_logged_language_redirection(self):
+        """Check that the customer is redirected to a page in his language."""
+        self.setup_website_customer_lang('en_US')
+        res = self.url_open(self.record_access_url)
+        self.assertNotIn(f'/en/my/website_test_portal/{self.record_portal.id}', res.url)
+        self.assertIn(f'/my/website_test_portal/{self.record_portal.id}', res.url)
+
+        for lang_code, expected_url_prefix in (('fr_FR', '/fr'), ('es_ES', '/es')):
+            lang = self.setup_website_customer_lang(lang_code)
+            # Language not enabled in the current website: no redirection to customer language to avoid 404
+            self.website.language_ids -= lang
+            res = self.url_open(self.record_access_url)
+            self.assertNotIn(f'{expected_url_prefix}/my/website_test_portal/{self.record_portal.id}', res.url)
+            self.assertIn(f'/my/website_test_portal/{self.record_portal.id}', res.url)
+            # Language enabled in the current website: redirection to customer language
+            self.website.language_ids += lang
+            res = self.url_open(self.record_access_url)
+            self.assertIn(f'{expected_url_prefix}/my/website_test_portal/{self.record_portal.id}', res.url)
+
+    @mute_logger('odoo.addons.base.models.ir_model')
+    def test_profile_language_redirection(self):
+        self.setup_website_customer_lang('es_ES')
```

Comment:
I would add some docstring to each testing case to explain the query differences.

Can be a subset of your big explanation of the PR :smile: